ast-checks/src/missing-patch-comment.rs

```diff
@@ -0,0 +1,99 @@
+use codespan::{FileId, Files};
+use comment_finders::{find_comment_above, find_comment_within};
+use rnix::{types::*, SyntaxNode};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, env, error::Error, fs};
+use tree_utils::walk_keyvalues_filter_key;
+
+mod comment_finders;
+mod tree_utils;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct SourceLocation {
+    column: usize,
+    line: usize,
+    file: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct MissingCommentMessage {
+    cond: bool,
+    locations: Vec<SourceLocation>,
+    msg: &'static str,
+    name: &'static str,
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let args: Vec<String> = env::args().skip(1).collect();
+    let mut report = HashMap::new();
+
+    let mut files = Files::new();
+    for filename in args {
+        let file_id = files.add(filename.clone(), fs::read_to_string(&filename)?);
+        report.insert(filename.clone(), execute(&files, file_id)?);
+    }
+
+    println!("{}", serde_json::to_string(&report)?);
+    Ok(())
+}
+
+fn execute(
+    files: &Files<String>,
+    file_id: FileId,
+) -> Result<Vec<MissingCommentMessage>, Box<dyn Error>> {
+    let ast = rnix::parse(files.source(file_id)).as_result()?;
+    let root = ast.root().inner().ok_or("No elements")?;
+    let mut report: Vec<MissingCommentMessage> = vec![];
+
+    let mut record_missing_patch_comment = |item: &SyntaxNode| -> Result<(), Box<dyn Error>> {
+        let start = item.text_range().start().to_usize() as u32;
+        let loc = files.location(file_id, start)?;
+
+        report.push(MissingCommentMessage {
+            cond: true,
+            msg: "Please add a comment on the line above explaining the purpose of this patch.",
+            name: "missing-patch-comment",
+            locations: vec![SourceLocation {
+                file: files
+                    .name(file_id)
+                    .to_str()
+                    .ok_or("encoding error")?
+                    .to_string(),
+                // 1-based to 0-based indexing
+                column: loc.column.to_usize() + 1,
+                line: loc.line.to_usize() + 1,
+            }],
+        });
+
+        Ok(())
+    };
+
+    // Find a key in an attrset called 'patches'
+    for kv in walk_keyvalues_filter_key(&root, "patches") {
+        // Look for a comment directly above the `patches = ...` line
+        if let Some(_comment) = find_comment_above(kv.node()) {
+            // Okay!, we found a comment directly above this line,
+            // so I'm interpreting that as the comment which applies to
+            // all patches. It would be _better_ to put a single comment
+            // per patch, but we'll let this slide.
+            continue;
+        }
+
+        // Try to convert the value to a list
+        if let Some(value) = kv.value().and_then(List::cast) {
```

Comment:
If this fails you just return the empty report? Seems like it would be worth throwing an error over?